### DISCUSSION TITLE
How to reconnect automatically (KeepAlive isn't working as I expect)

### DISCUSSION THREAD

**Discussion by Podoproas** [FreeOpcUa/python-opcua#1389] [Q&A]
Hi everyone,
I'm new to using OPC-UA technology and I am interested in how to make a persistent connection.
I have a client, that subscribes on some server and retrieve some data from that - after that it stores the data in DB. I want to do that even if Server was turned off for some time, but it doesn't.

As I understood, Client.connect(url) call create_session method that uses KeepAlive class for renewing secure_channel in a new thread. But as I did not expect any socket Error can stop KeepAlive thread execution:

_client.py_:
```
try:
    self.client.open_secure_channel(renew=True)
except concurrent.futures.TimeoutError:
    _logger.debug("keepalive failed: timeout on open_secure_channel()")
    break
```

 So secure_channel will be lost, and data will never be retrieved for this client anymore. What should I do? 
How to know that client has disconnected? How should I restart it properly? Is there a better way to handle such a situation?

To clarify my intentions:
I want to have a running Server instance, that has method "add_client(ip)" - after that client instance is created and subscribed on the other (PLC local) Server nodes change.(I want to have different clients for different PLC OPC-UA servers) 

This client instance makes it possible to store the data in the DB. It is obvious that PLC Server may experience some problems such as rebooting, lack of internet, e.t.c. But after these problems solved I want to receive data anyway.

**Comment by AndreasHeine** [comment]
If the Server went off the Session is gone there is no persistent Session!

```
To clarify my intentions:
I want to have a running Server instance, that has method "add_client(ip)" - after that client instance is created and subscribed on the other (PLC local) Server nodes change.(I want to have different clients for different PLC OPC-UA servers)
```
that sounds like the FindServers and FindServersOnNetwork discovery part?
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.4.2/

**Reply by Podoproas** [reply]
> there is no persistent Session!

The only question I now have is "Why?"
Do you think it is useless? May be I could do it and make a PR?

**Reply by AndreasHeine** [reply]
its not a question of usefull or not, we need to check the opc ua spec. first i am not sure that this is specified!? https://reference.opcfoundation.org/v104/Core/docs/Part4/6.7/

**Reply by oroulet** [reply]
Persistent Session is quite a lot of work to implement. Send a good PR and i will merge
